File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        
        ListNode *cur = head;
        int len = 0;
        while(cur != NULL)
        {
            cur = cur->next;
            len++;
        }
        int rem = len - n;
        if(rem == 0)
        {
            head = head->next;
            return head;
        }
        else
        {
            int i =0;
            ListNode *temp = head;
            while(i != rem-1)
            {
                temp = temp->next;
                i++;
            }
            ListNode *ptr = temp->next->next;
            temp->next = ptr;
        }
        
        return head;
        
        
        
    }
};

File: 19-remove-nth-node-from-end-of-list/19-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        ListNode p1= head;
        ListNode p2 = head;
        ListNode prev = null;
        int c = 0;
        while(c < n && p2!=null )
        {
            prev = p2;
            p2 = p2.next;
            c++;
        }
       // System.out.println(p2.val);
        if(c == n && p2 == null)
        {
            head = head.next;
            return head;
        }
        if(p2 == null)
            return null;
        
        
        while(p2.next != null)
        {
            p1 = p1.next;
            p2 = p2.next;
        }
        p1.next = p1.next.next;
        
        return head;
    }
}